### DISCUSSION TITLE
Servicebus queues using CARML modules

### DISCUSSION THREAD

**Discussion by ca55ar** [Azure/ResourceModules#1399] [Q&A] [answered]
Hi, can anyone give me a working example of how I can create multiple servicebus queues in multiple servicebus namespaces using the CARML module library?

The documentation [here ](https://github.com/Azure/ResourceModules/tree/main/arm/Microsoft.ServiceBus/namespaces)seems to say that when creating a servicebus namespace, I can use "queues" as a parameter and give it an array to create a bunch of queues in that namespace.

When I do this using these 2 files (rgs is the resources groups i am successfully creating prior to the servicebus namespaces):
main.bicep
```

param servicebusParams array

// creates all the Service Bus' in to the Resource Groups in the scope provided
module sbs '../ResourceModules-0.5.0/arm/Microsoft.ServiceBus/namespaces/deploy.bicep' = [for sbparam in servicebusParams: {
  name: '${sbparam.Name}-deploy'
  scope: resourceGroup(sbparam.resourceGroupToDeployTo)
  params: {
    name: sbparam.name
    location: sbparam.location
    queues: sbparam.queues
  }
  dependsOn: [
    rgs
  ]
}]
```

with this as the params.json:
```

"servicebusParams": {
            "value": [
                {
                    "name": "sb-fruits",
                    "location": "northeurope",
                    "resourceGroupToDeployTo": "rg-1",
                    "skuName": "Basic",
                    "queues": {
                        "Value": [
                            "banana",
                            "apple",
                            "jackfruit"
                        ]
                    }
                },
                {
                    "name": "sb-bikes",
                    "location": "northeurope",
                    "resourceGroupToDeployTo": "rg-2",
                    "skuName": "Basic",
                     "queues": {
                        "Value": [
                            "harley",
                            "ducati",
                            "ninja"
                        ]
                    }
                }
            ]
        }
```

... I get an error saying that is expecting a type 'Array' but received a value of type 'Object'
Inner Errors:
{'code': 'InvalidTemplate', 'target': '/subscriptions/b15e52e6-e278-4362-bf92-f6514a5b7f36/resourceGroups/rg-1/providers/Microsoft.Resources/deployments/sb-fruits-deploy', 'message': "Deployment template validation failed: 'Template parameter 'queues' was provided an invalid value. Expected a value of type 'Array', but received a value of type 'Object'. Please see https://aka.ms/resource-manager-parameter-files for usage details.'.", 'additionalInfo': [{'type': 'TemplateViolation', 'info': {'lineNumber': 1, 'linePosition': 5018, 'path': 'properties.template.parameters.queues'}}]}

So I thought it may be the way the main.bicep file is calling the queues param from the json file, so I tried with this instead, setting the queues param as a array:

```
// creates all the Service Bus' in to the Resource Groups in the scope provided
module sbs '../ResourceModules-0.5.0/arm/Microsoft.ServiceBus/namespaces/deploy.bicep' = [for sbparam in servicebusParams: {
  name: '${sbparam.Name}-deploy'
  scope: resourceGroup(sbparam.resourceGroupToDeployTo)
  params: {
    name: sbparam.name
    location: sbparam.location
    queues: [
      sbparam.queues
    ]
    }
  dependsOn: [
    rgs
  ]
}]
```
and then i get this error:
\"message\": \"Unable to process template language expressions for resource '/subscriptions/b15e52e6-e278-4362-bf92-f6514a5b7f36/resourceGroups/rg-1/providers/Microsoft.Resources/deployments/xiwar6wodsuba-Queue-0' at line '1' and column '26213'. 'The language expression property 'name' doesn't exist, available properties are 'Value'.

Am I doing this wrong? do I need to call the queues module separately? Is my syntax messed up?
It just seems like from the documentation on the servicebus namespace site, it should work.

Cheers

**Comment by ca55ar** [comment, marked as answer]
Hi [alex-frankel](https://github.com/alex-frankel)  [AlexanderSher](https://github.com/AlexanderSher) [brwilkinson](https://github.com/brwilkinson) Is there any help available for CARML? Just wondering if it is worth putting in the time to use it if there isn't much support about for it. Would be a shame. I am prob just not doing something right, as I am still new to it all.

I am seeing a similar issues with creating key vault secrets and key vault accessPolicies. So I think I am struggling with creating child resources generally. The docs seem to say I can call these parameters from the parent module, but I don't seem to be able to. I have been able to create the keyvault secrets by doing a separate module call to the '../ResourceModules-0.5.0/arm/Microsoft.KeyVault/vaults/secrets/deploy.bicep' but still having trouble with the access policies. Would be helpful to have examples of how to use these.

Thanks again, and if I don't hear anything in a day or 2, I will assume this one will drift in to the ether.

**Reply by ca55ar** [reply]
I have been able to get most stuff going now. There are quite a few areas where is says in the docs that you can call child modules from the main module file, and it should work, but in my experience, that has not been the case. I have to call those individually as modules in their own right in my main.bicep file. But that's ok.

**Reply by AlexanderSehr** [reply]
Just noticed why this notification didn't show up - it's actually @MrMCake ;) Was never able to rename my GitHub handle.

**Comment by AlexanderSehr** [comment]
Hey @ca55ar , sorry for the late reply. Not sure what to make of GitHub's `discussions` feature as I somehow seem to miss notifications if there is a new one. 

Anyhow, to your question, if it is still relevant: You should absolutely be able to deploy child-modules via there parent. I hope we're talking about the same thing though. What we intend to express is, that you can deploy a KeyVault with AccessPolicies & Secrets using for example the following syntax:

```Bicep
module testDeployment '../../deploy.bicep' = {
  scope: resourceGroup
  name: '${uniqueString(deployment().name)}-test-${serviceShort}'
  params: {
    name: '<<namePrefix>>${serviceShort}001'
    accessPolicies: [
      {
        objectId: resourceGroupResources.outputs.managedIdentityPrincipalId
        permissions: {
          keys: [
            'get'
            'list'
            'update'
          ]
          secrets: [
            'all'
          ]
        }
        tenantId: tenant().tenantId
      }
      {
        objectId: resourceGroupResources.outputs.managedIdentityPrincipalId
        permissions: {
          certificates: [
            'backup'
            'create'
            'delete'
          ]
          secrets: [
            'all'
          ]
        }
      }
    ]
    secrets: {
      secureList: [
        {
          attributesExp: 1702648632
          attributesNbf: 10000
          contentType: 'Something'
          name: 'secretName'
          roleAssignments: [
            {
              principalIds: [
                resourceGroupResources.outputs.managedIdentityPrincipalId
              ]
              roleDefinitionIdOrName: 'Reader'
            }
          ]
          value: 'secretValue'
        }
      ]
    }
  }
}
```

For all our modules, we run usually multiple tests that are designed to test as many aspects of a module as we possibly can. In that sense it's always worth checking the module's `readme.md` or the `.test` folder as both provide you with examples as per the tests we're running.

I hope this provided some insights, and again sorry for the late reply. The repository is very active and issues weekly reviewed. I'll see to highlight the `Discussions` in the future too.

**Comment by AlexanderSehr** [comment]
Regarding the Service Bus:

I see an syntactical error in the provided parameters block which likely was the reason for the original error. To recap, you provided the following JSON example

```json
"servicebusParams": {
    "value": [
        {
            "name": "sb-fruits",
            "location": "northeurope",
            "resourceGroupToDeployTo": "rg-1",
            "skuName": "Basic",
            "queues": {
                "Value": [
                    "banana",
                    "apple",
                    "jackfruit"
                ]
            }
        },
        {
            "name": "sb-bikes",
            "location": "northeurope",
            "resourceGroupToDeployTo": "rg-2",
            "skuName": "Basic",
            "queues": {
                "Value": [
                    "harley",
                    "ducati",
                    "ninja"
                ]
            }
        }
    ]
}
```

The issue here is that the currently common parameter files, only expect a `value` specification for any top-level parameter (i.e., the ones that exist right underneath? the `"parameters" {}` block.

That means, you can remove any 'inner' `value` reference and the parameter file should look like:

```json
"servicebusParams": {
    "value": [
        {
            "name": "sb-fruits",
            "location": "northeurope",
            "resourceGroupToDeployTo": "rg-1",
            "skuName": "Basic",
            "queues": [
                "banana",
                "apple",
                "jackfruit"
            ]
        },
        {
            "name": "sb-bikes",
            "location": "northeurope",
            "resourceGroupToDeployTo": "rg-2",
            "skuName": "Basic",
            "queues": [
                "harley",
                "ducati",
                "ninja"
            ]
        }
    ]
}
```
> Note the changes of the `queue` param
